arm/examples/sdk.go

```diff
@@ -0,0 +1,168 @@
+
+import (
+	"fmt"
+	"os"
+
+	"bytes"
+	"encoding/json"
+	"github.com/Azure/azure-sdk-for-go/arm/examples/helpers"
+	"github.com/Azure/azure-sdk-for-go/arm/resources/resources"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure"
+)
+
+// SDK represents a working authenticated SDK structure.
+type SDK struct {
+
+	// Authentication holds misc. authentication information for interacting with the SDK.
+	Authentication *Authentication
+}
+
+// Authentication represents the values needed to authenticate with the ARM API.
+type Authentication struct {
+
+	// ClientID is the `appId` from an Azure Service Principal.
+	ClientID string
+
+	// ClientSecret is the `password` from an Azure Service Principal.
+	ClientSecret string
+
+	// SubscriptionID is the unique Azure subscription ID to use.
+	SubscriptionID string
+
+	// TenantID is the `tenant` from an Azure Service Principal.
+	TenantID string
+
+	// HashMap is the data value some parts of the SDK will expect, and is stored here as convenience.
+	HashMap map[string]string
+
+	// AuthenticatedToken is the generated token that is needed to authenticate with each client.
+	AuthenticatedToken *adal.ServicePrincipalToken
+}
+
+func NewSDK() (*SDK, error) {
+
+	// -------------------------------------------------------------------
+	// Validation that the environment is configured correctly.
+	//
+	clientID := os.Getenv("AZURE_CLIENT_ID")
+	if clientID == "" {
+		return nil, fmt.Errorf("empty $AZURE_CLIENT_ID")
+	}
+	clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
+	if clientSecret == "" {
+		return nil, fmt.Errorf("empty $AZURE_CLIENT_SECRET")
+	}
+	subscriptionID := os.Getenv("AZURE_SUBSCRIPTION_ID")
+	if subscriptionID == "" {
+		return nil, fmt.Errorf("empty $AZURE_SUBSCRIPTION_ID")
+	}
+	tenantID := os.Getenv("AZURE_TENANT_ID")
+	if tenantID == "" {
+		return nil, fmt.Errorf("empty $AZURE_TENANT_ID")
+	}
+
+	// Here we build the SDK based on a user's configuration.
+	sdk := &SDK{
+		Authentication: &Authentication{
+			ClientID:       clientID,
+			ClientSecret:   clientSecret,
+			SubscriptionID: subscriptionID,
+			TenantID:       tenantID,
+			HashMap: map[string]string{
+				"AZURE_CLIENT_ID":       clientID,
+				"AZURE_CLIENT_SECRET":   clientSecret,
+				"AZURE_SUBSCRIPTION_ID": subscriptionID,
+				"AZURE_TENANT_ID":       tenantID,
+			},
+		},
+	}
+
+	// Generate an authenticated token.
+	authenticatedToken, err := helpers.NewServicePrincipalTokenFromCredentials(sdk.Authentication.HashMap, azure.PublicCloud.ResourceManagerEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	sdk.Authentication.AuthenticatedToken = authenticatedToken
+	return sdk, nil
+}
+
+// E will check if an error is nil, and exit the program in the case of an error.
+// Use this as convenience to simplify code in various examples.
+func E(err error) {
+	if err != nil {
+		fmt.Printf("Error: %v\n", err)
+		os.Exit(2)
```

Comment:
I'd rather not `os.Exit` so that we can give the program one last chance to run any `defer` statements. As an example, that may include deleting a resource group that was created to contain the assets created by this sample. (As alluded to in another comment)